packages/botonic-core/src/core-bot.ts

```diff
@@ -91,194 +91,149 @@ export class CoreBot {
+    await this.runPostPlugins(botContext, response)
 
-    session.is_first_interaction = false
+    botContext.session.is_first_interaction = false
 
     return {
-      input,
+      input: botContext.input,
       response,
-      session,
-      lastRoutePath,
+      session: botContext.session,
+      lastRoutePath: botContext.lastRoutePath,
     }
   }
 
-  private async runPrePlugins(
-    input: Input,
-    session: Session,
-    lastRoutePath: RoutePath
-  ) {
+  private async runPrePlugins(botContext: BotContext) {
     if (this.plugins) {
-      await runPlugins(
-        this.plugins,
-        'pre',
-        input,
-        session,
-        lastRoutePath,
-        undefined
-      )
+      await runPlugins({ botContext, mode: 'pre' })
     }
   }
 
-  private async getOutput(
-    input: Input,
-    session: Session,
-    lastRoutePath: RoutePath
-  ): Promise<ProcessInputResult> {
-    await this.setRouter(input, session, lastRoutePath)
+  private async getOutput(botContext: BotContext): Promise<ProcessInputResult> {
+    await this.setRouter(botContext)
 
     const output = (this.router as Router).processInput(
-      input,
-      session,
-      lastRoutePath
+      botContext.input,
+      botContext.session,
+      botContext.lastRoutePath
     )
     return output
   }
 
-  private async setRouter(
-    input: Input,
-    session: Session,
-    lastRoutePath: RoutePath
-  ) {
+  private async setRouter(botContext: BotContext) {
     if (this.routes instanceof Function) {
       this.router = new Router(
         [
-          ...(await getComputedRoutes(this.routes, {
-            input,
-            session,
-            lastRoutePath,
-          })),
+          ...(await getComputedRoutes(this.routes, botContext)),
           ...this.defaultRoutes,
         ],
         this.inspector.routeInspector
       )
     }
   }
 
-  private createRequestFromOutput(
-    { input, session, lastRoutePath }: BotRequest,
+  private updateRequestFromOutput(
+    botContext: BotContext,
     output: ProcessInputResult
-  ) {
+  ): BotContext {
     return {
-      getString: (stringId: string) => this.getString(stringId, session),
-      setLocale: (locale: string) => this.setLocale(locale, session),
-      session: session || {},
+      ...botContext,
       params: output.params || {},
-      input,
-      plugins: this.plugins,
-      defaultTyping: this.defaultTyping,
-      defaultDelay: this.defaultDelay,
-      lastRoutePath,
+      lastRoutePath: output.lastRoutePath,
     }
   }
 
-  private async runPostPlugins(
-    input: Input,
-    session: Session,
-    lastRoutePath: RoutePath,
-    response: any
-  ) {
+  private async runPostPlugins(botContext: BotContext, response: any) {
     if (this.plugins) {
-      await runPlugins(
-        this.plugins,
-        'post',
-        input,
-        session,
-        lastRoutePath,
-        response
-      )
+      await runPlugins({
+        botContext,
+        mode: 'post',
```

Comment:
nit: it would be nice to have type/enum for this `PluginMode.Pre|Post`

Reply:
I'll apply it in another PR